startup: Make "-s -" read from stdin

ZyX-I opened the pull request.

ZyX-I committed:
main: Make `-s -` read from stdin

ZyX-I committed:
functests: Check that `-s` works as expected

ZyX-I commented on test/functional/core/main_spec.lua:
```diff
@@ -0,0 +1,60 @@
+      end
+      neq(0, max_sec)
+    end)
+    it('does not expand $VAR', function()
```
This test is there because `-S` is not working well with such names (`-S arg` uses a concat of `source` and `arg` without any escaping used). `-s` is treated different and does not have the problem, but this option looks similar to `-S`.

ZyX-I committed:
tests: Fix CI failures

ZyX-I committed:
tests: Fix repeated_popen_r usage, rename the function

ZyX-I committed:
functests: Alter the order of checks

Check whether `repeated_read_cmd` returned nil and did not actually run nvim or 
it did, but still returned nil for whatever reason.

ZyX-I committed:
functests: Make sure that line ending is LF and not CRLF

ZyX-I committed:
functests: Use Neovim instance and system() in place of lua io.popen

ZyX-I committed:
functests: Fix testlint errors

ZyX-I committed:
functests: Do not check stdout, it is different on Windows

oni-link commented:
With `nvim -u NONE -s -` and `:qa` I get this:

```bash
nvim: /home/oni-link/git/neovim/.deps/build/src/libuv/src/unix/core.c:521: uv__close: Assertion `fd > STDERR_FILENO' failed.
Aborted
```

ZyX-I commented:
@oni-link Interesting, I can reproduce this in tmux. I can’t reproduce it when running in Neovim terminal with exactly the same arguments (`nvim_prog_abs(), '-u', 'NONE', '-i', 'NONE', '--cmd', 'set noswapfile shortmess+=IFW fileformats=unix', '-s', '-'`). Note: it crashes on `:`, before I can type `q`.

Solution is obvious currently: use `dup()` to clone `STDIN_FILENO` and use the resulting fd. Though will need to use os/fileio.c in place of `f*` functions: `fdopen` is not in C99.

ZyX-I commented:
A bit different: I can reproduce this in `termopen()`, but either if I run it in an interactive shell or if I actually close inner Neovim, in the last case problem appears not when typing `:`, but during shutdown.

ZyX-I committed:
main: Temporary fix assertion error

This variant uses `fdopen()` which is not standard, but it fixes problem on my 
system. In next commit `scriptin` will use `FileDescriptor*` from os/fileio in 
place of `FILE*`.

ZyX-I committed:
getchar: Use fileio instead of fdopen

Problem: as fileio is cached and reads blocks this is going to wait
until either EOF or reading enough characters to fill rbuffer. This is
not good when reading user input from stdin as script.

ZyX-I committed:
fileio,main: Do not restart syscall at EAGAIN when reading for -s

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
os/fileio: Fix QB failure

ZyX-I committed:
functests: Test -s errors

ZyX-I committed:
unittests: Add tests for new fileio functions

ZyX-I committed:
main: Translate full -s error message, not part of it

justinmk commented:
Windows build hangs, `termopen()` won't work there, yet.

justinmk commented on test/functional/core/main_spec.lua:
```diff
@@ -0,0 +1,130 @@
+      local attrs = lfs.attributes(fname)
+      eq(#('100500\n'), attrs.size)
+    end)
+    it('does not crash after reading from stdin in non-headless mode', function()
```
Add 

    if helpers.pending_win32(pending) then return end

here, since this test uses termopen().

justinmk replied:
termopen() works on Windows now, so the above comment is outdated. Though reading from stdin on Windows may be broken (https://github.com/neovim/neovim/issues/6890) so maybe it's expedient to  skip it for now.

ZyX-I commented:
It hangs on the very first test, and that does not use `termopen()`, it uses `system()`.

ZyX-I committed:
functests: Do not run termopen test on Windows

ZyX-I committed:
functests: Disable system(-s -) test on Windows

Assume something with system() if second test hangs as well. Assume something 
with reading stdin if not.

justinmk commented on src/nvim/os/fileio.c:
```diff
@@ -69,14 +68,45 @@ int file_open(FileDescriptor *const ret_fp, const char *const fname,
+/// @param[in]  fd  File descriptor to wrap.
+/// @param[in]  flags  Flags, @see FileOpenFlags. Currently reading from and
+///                    writing to the file at once is not supported, so either
+///                    FILE_WRITE_ONLY or FILE_READ_ONLY is required.
```
should this be:

    kFileWriteOnly or kFileReadOnly is required.

justinmk commented on src/nvim/os/fileio.h:
```diff
@@ -32,6 +33,8 @@ typedef enum {
   kFileAppend = 64,  ///< Append to the file. Implies kFileWriteOnly. Cannot
                      ///< be used with kFileCreateOnly.
+  kFileNonBlocking = 128,  ///< Do not restart read() or write() syscall if
+                           ///< EAGAIN was encountered.
```
"EAGAIN or EINTR" ?

ZyX-I replied:
EAGAIN only.

ZyX-I committed:
Merge branch 'master' into s-dash-stdin

ZyX-I committed:
getchar: Move REMAP_… values into a enum

ZyX-I commented:
Looks like 

```
[ RUN      ] Command-line option -s does not expand $VAR: 
```

hang again.

ZyX-I commented:
If I recall correctly that is “using `-s` always hangs” rather then something specific to `$fname`.

justinmk commented:
Is it a regression or is it just a new test that can be skipped for now?

ZyX-I commented:
It is a new test, but I am not aware of the similar test that checked something before .

BTW, is there some script to build Neovim for wine? E.g. like I did for Vim once: https://bitbucket.org/ZyX_I/vim-dev-tools/src/342d10f67bbbb072288b80d0c7effb2507efe21b/winesetup.pl?at=default prepares WINEPREFIX (also builds Vim), specifically build.zsh near that uses mingw to build vim.

justinmk commented:
> BTW, is there some script to build Neovim for wine?

Not that I know.

ZyX-I commented:
I started working on that, but soon found luarocks/luarocks#749. Most likely that means that quick “just make luarocks compile things for me” is out and all needed packages would be cross-compiled, or alternatives (like [lullpeg](https://github.com/pygy/LuLPeg)) found, though I will try to determine why `os.execute` is incompatible on wine to that extent. Currently I can say that

   luajit -e "print(os.execute(\"\\\"C:\\luarocks\\tools\\mkdir\\\" -p \\\"abc\\\"\"))"

fails, but succeeds without one pair of quotes. Reported two bugs to wine as a result: https://bugs.winehq.org/show_bug.cgi?id=44133 and https://bugs.winehq.org/show_bug.cgi?id=44132.

justinmk commented:
@ZyX-I If you have a RDP client you can connect to appveyor for 1-2 hours, useful for sanity checks. Requires a temporary change to appveyor.yml, a bit annoying but it works well. https://www.appveyor.com/docs/how-to/rdp-to-build-worker/

justinmk commented:
Let's just ignore that test on Windows:

    Command-line option -s does not expand $VAR: 

Work is being done to improve the tests on Windows in https://github.com/neovim/neovim/pull/7412

justinmk commented:
After ignoring the "does not expand $VAR" test in https://github.com/neovim/neovim/pull/7713 now the "old tests" hang (since https://github.com/neovim/neovim/pull/7705 , the Vim "old tests" run on AppVeyor).

This hangs:

```
VIMRUNTIME=../../../runtime; export VIMRUNTIME;  ../../../build/bin/nvim.exe -u unix.vim -U NONE -i viminfo --headless --noplugin -s dotest.in test1.in
```

So there is a regression in the behavior of `-s` on Windows.

justinmk commented on src/nvim/main.c:
```diff
@@ -1098,17 +1099,25 @@ static void command_line_scan(mparm_T *parmp)
-              mch_errmsg("\"\n");
+            int error;
+            if (STRCMP(argv[0], "-") == 0) {
+              const int stdin_dup_fd = os_dup(OS_STDIN_FILENO);
```
@ZyX-I maybe https://github.com/neovim/neovim/pull/8267 is helpful to implement `os_dup` on Windows.

justinmk commented on test/functional/core/main_spec.lua:
```diff
@@ -0,0 +1,132 @@
+  end
+end
+
+describe('Command-line option', function()
```
this could go in `startup_spec.lua`
